[Samrt Accounts Kit][Troubleshooting Section] Add invalid signature (#2768)
MetaMask/metamask-docs: smart-accounts-kit/troubleshooting/invalid-signature.md

---
description: How to resolve the InvalidEOASignature error when redeeming delegations.
sidebar_label: Invalid EOA signature
keywords: [InvalidEOASignature, error code, delegation, troubleshooting, invalid EOA signature]
---

# Invalid EOA signature

The Delegation Manager reverts with `InvalidEOASignature()` in the following cases.

## Smart account is not deployed

The root delegation's delegator must be a MetaMask smart account. The Delegation Manager
checks delegator code length to determine whether the delegator is an EOA or a smart
account. 

If the smart account hasn't been deployed yet, it has no code at its address,
so the Delegation Manager treats it as an EOA and attempts ECDSA signature recovery.
Since the delegation was signed by the smart account, recovery produces a different address and reverts.

### Solution

Verify the delegator's smart account is deployed before the delegation is redeemed. The
smart account can be either an ERC-4337 smart account or an EIP-7702 upgraded EOA. 

For an ERC-4337 smart account, the first user operation sent from the account automatically

> deploys it. See [deploy a smart account guide](../guides/smart-accounts/deploy-smart-account.md) to learn more

Reviewer: ```suggestion
For an ERC-4337 smart account, the first user operation sent from that account deploys it
automatically. For more information, see the [deploy a smart account guide](../guides/smart-accounts/deploy-smart-account.md) guide.
```

For an EIP-7702 upgraded EOA, verify that the authorization to set the

> account code has been submitted before the delegation is redeemed.

Reviewer: ```suggestion
For an EIP-7702-upgraded EOA, verify that you submit the authorization to set the account code
before redeeming the delegation.
```

```ts
import { getSmartAccountsEnvironment } from "@metamask/smart-accounts-kit";
import { sepolia as chain } from "viem/chains";

// Get the EOA account code
const code = await publicClient.getCode({
  address,
});

if (code) {
  // According to EIP-7702, the code format is 0xef0100 || address.
  // Remove the first 8 characters (0xef0100) to get the delegator address.
  const delegatorAddress = `0x${code.substring(8)}`;

  const statelessDelegatorAddress = getSmartAccountsEnvironment(chain.id)
  .implementations
  .EIP7702StatelessDeleGatorImpl;

  // If the account isn't upgraded to a MetaMask smart account, you can
  // either upgrade programmatically or ask the user to switch to a smart account manually.
  const isAccountUpgraded = delegatorAddress.toLowerCase() === statelessDelegatorAddress.toLowerCase();
}
```

> If the EOA isn't upgraded to a MetaMask smart account, see the [EIP-7702 quickstart guide](../get-started/smart-account-quickstart/eip7702.md).

Reviewer: ```suggestion
To upgrade an EOA to a MetaMask smart account, see the [EIP-7702 quickstart guide](../get-started/smart-account-quickstart/eip7702.md).
```

## Incorrect signer

The delegation was signed with an account that doesn't correspond to the delegator
address. When the delegator is an EOA, the Delegation Manager recovers the signer from
the EIP-712 typed data hash and compares it to the `delegator` field. If they don't
match, the transaction reverts.

> This case occurs when redeeming a [delegation chain](../guides/delegation/create-redelegation.md), where an intermediate or leaf delegation has an EOA as the delegator and the delegation is signed with the wrong account.

Reviewer: ```suggestion
This occurs when redeeming a [delegation chain](../guides/delegation/create-redelegation.md). An
intermediate or leaf delegation has an EOA as the delegator, but the delegation is signed by an
account other than the expected delegate.
```
It always does my head in when we have "delegation", "delegator", and "delegate" in the same paragraph.
